### PR TITLE
haskellPackages: allow hnix-store-{core,remote} 0.5

### PR DESCRIPTION
**Pull request by Anton-Latukha**
These were blocked due to mutual desigion during me<>`sternenseemann`
discussion.

In short:

I shipped my own work (to support GHC 9.0) in the 0.5 releases of
`hnix-store-{core,remote}`.

These packages are really used only by `hnix` itself, and instead of maintaining
them in Nixpkgs & reacting on `hnix` release, we decided to hold them back &
switch to these versions when `hnix` provides support for them.

I just (today) released `hnix` 0.14 & it requires `hnix-store-{core,remote}`
0.5.

If you would look at dependency tree:
https://packdeps.haskellers.com/reverse/hnix-store-core
https://packdeps.haskellers.com/reverse/hnix-store-remote

It shows that `hnix` currently is the only alive use of these projects.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sternenseemann** [comment]
If you run `./maintainers/scripts/haskell/regenerate-hackage-packages.sh` you should be able to see the effects of your changes and test them as well (it generates `pkgs/development/haskell-modules/hackage-packages.nix`).

**Comment by Anton-Latukha** [comment]
Recieved a clean diff:
```diff
diff --git a/pkgs/development/haskell-modules/hackage-packages.nix b/pkgs/development/haskell-modules/hackage-packages.nix
index 364581829a7..a00deb16fff 100644
--- a/pkgs/development/haskell-modules/hackage-packages.nix
+++ b/pkgs/development/haskell-modules/hackage-packages.nix
@@ -131462,41 +131462,10 @@ self: {
        ];
        description = "Haskell implementation of the Nix language";
        license = lib.licenses.bsd3;
+       maintainers = with lib.maintainers; [ Anton-Latukha ];
      }) {};
 
   "hnix-store-core" = callPackage
-    ({ mkDerivation, algebraic-graphs, attoparsec, base
-     , base16-bytestring, base64-bytestring, binary, bytestring, cereal
-     , containers, cryptohash-md5, cryptohash-sha1, cryptohash-sha256
-     , cryptohash-sha512, directory, filepath, hashable, hspec
-     , lifted-base, monad-control, mtl, nix-derivation, process, saltine
-     , tasty, tasty-discover, tasty-golden, tasty-hspec, tasty-hunit
-     , tasty-quickcheck, temporary, text, time, unix
-     , unordered-containers, vector
-     }:
-     mkDerivation {
-       pname = "hnix-store-core";
-       version = "0.4.3.0";
-       sha256 = "17l7198s3dzjx416nplinyghbqgignvpvxkww241gjv4p23izf77";
-       libraryHaskellDepends = [
-         algebraic-graphs attoparsec base base16-bytestring
-         base64-bytestring bytestring cereal containers cryptohash-md5
-         cryptohash-sha1 cryptohash-sha256 cryptohash-sha512 directory
-         filepath hashable lifted-base monad-control mtl nix-derivation
-         saltine text time unix unordered-containers vector
-       ];
-       testHaskellDepends = [
-         attoparsec base base16-bytestring base64-bytestring binary
-         bytestring containers directory filepath hspec process tasty
-         tasty-golden tasty-hspec tasty-hunit tasty-quickcheck temporary
-         text unix
-       ];
-       testToolDepends = [ tasty-discover ];
-       description = "Core effects for interacting with the Nix store";
-       license = lib.licenses.asl20;
-     }) {};
-
-  "hnix-store-core_0_5_0_0" = callPackage
     ({ mkDerivation, algebraic-graphs, attoparsec, base
      , base16-bytestring, base64-bytestring, binary, bytestring, cereal
      , containers, cryptonite, directory, filepath, hashable, hspec
@@ -131526,27 +131495,10 @@ self: {
        testToolDepends = [ tasty-discover ];
        description = "Core effects for interacting with the Nix store";
        license = lib.licenses.asl20;
-       hydraPlatforms = lib.platforms.none;
+       maintainers = with lib.maintainers; [ Anton-Latukha ];
      }) {};
 
   "hnix-store-remote" = callPackage
-    ({ mkDerivation, attoparsec, base, binary, bytestring, containers
-     , hnix-store-core, mtl, network, nix-derivation, text, time
-     , unordered-containers
-     }:
-     mkDerivation {
-       pname = "hnix-store-remote";
-       version = "0.4.3.1";
-       sha256 = "1wp6z5f13fkgh3mdzr1zcqdvsx1j90l5isf671gsg43brsv4c2vl";
-       libraryHaskellDepends = [
-         attoparsec base binary bytestring containers hnix-store-core mtl
-         network nix-derivation text time unordered-containers
-       ];
-       description = "Remote hnix store";
-       license = lib.licenses.asl20;
-     }) {};
-
-  "hnix-store-remote_0_5_0_0" = callPackage
     ({ mkDerivation, attoparsec, base, binary, bytestring, containers
      , cryptonite, hnix-store-core, mtl, network, nix-derivation, text
      , time, unordered-containers
@@ -131562,7 +131514,7 @@ self: {
        ];
        description = "Remote hnix store";
        license = lib.licenses.asl20;
-       hydraPlatforms = lib.platforms.none;
+       maintainers = with lib.maintainers; [ Anton-Latukha ];
      }) {};
 
   "hnn" = callPackage
```

**Comment by sternenseemann** [comment]
Feel free to commit this as well here (or right before unmarking as WIP to avoid merge conflicts).

**Comment by Anton-Latukha** [comment]
Just an idea: if to produce a lot of files (maybe a file per project/package) - there probably would be drastically fewer regeneration diff conflicts (I think).

**Comment by sternenseemann** [comment]
@ofborg build haskellPackages.hnix

**Comment by sternenseemann** [comment]
When was the hnix release supporting the store packages?

**Comment by maralorn** [comment]
> Just an idea: if to produce a lot of files (maybe a file per project/package) - there probably would be drastically fewer regeneration diff conflicts (I think).

I don‘t think this is necessary. Merge conflicts in the hackage-packages.nix are not that common in the first place.
I have never noticed a merge conflict caused by two unrelated packages. I assume the 99% of all our merge conflicts come from changes to the same package. And those are honest, real merge conflicts.

**Comment by Anton-Latukha** [comment]
> When was the hnix release supporting the store packages?

Ttoday. After releases, if I can I contribute patches to Nixpkgs.

**Comment by Gabriel439** [comment]
I believe this is safe to merge into `haskell-updates` as the latest version of `hnix` on Hackage supports the latest versions of `hnix-store-{core,remote}`.  I can merge this myself tomorrow if nobody objects

**Comment by sternenseemann** [comment]
I guess we'll need to decide what to do about our hnix patches as well. The commits I originally reverted no longer apply with 0.14.*, so we either need update these patches and vendor them in or bite the bullet and override `relude_1_0_0_1` and `semialign_1_2`, but I have no idea if this causes any issues in reverse dependencies…

**Comment by sternenseemann** [comment]
@ofborg build haskellPackages.nix-linter haskellPackages.dhall-nix haskellPackages.dhall-nixpkgs haskellPackages.yarn2nix

(`update-nix-fetchgit` builds locally, but its test suite fails, I guess that is something @expipiplus1 can have a look at later.)

**Review by sternenseemann — APPROVED** [review]

**Pull request merged**
